### ISSUE TITLE
[RFC] Plugins

### ISSUE THREAD

**Issue by timneutkens** [vercel/next.js#9133]
Authored together with @ijjk.

## Goals

- Provide an official plugin system for Next.js
- Allow hooking into different parts of the Next.js lifecycle
- Provide the right limits for plugins to prevent them from de-optimizing applications
- Warn users when a plugin is being inefficient
- Provide a zero-config approach to plugins
- Provide code splitting for plugins

## Background

Over time the Next.js examples directory has grown to 210+ examples. These examples show how to integrate external libraries for styling, error reporting, data fetching, and many other applications.

Most of these examples require a bit of configuration in various parts of Next.js, for example, in `_document`, `_app`, or pages. And sometimes the implementation details have to match exactly the example otherwise the library behaves in a different way. We've seen this for the example with material-ui that has a flash of unstyled content if `_document` is not implemented.

There are also examples that are using `getInitialProps` where it's not always needed, one of the examples of this is the redux example. It currently requires having `getInitialProps` which causes all pages to opt-out of automatic static rendering.

To solve these issues and provide an overall better developer experience we intend to introduce a plugin system.

Plugins will be able to hook into specific lifecycle parts of Next.js.

For example, to execute a specific bit of code after hydration happens or to inject a script like Google Analytics into the body of every page.

The main goal for the plugin system will be to make you more productive while preserving the Next.js production performance characteristics.

One way we're trying to achieve this is that attached lifecycle methods are automatically code-split in a similar fashion as pages.

This new plugins system will provide you with the best developer experience when integrating libraries and we're excited to see the new ways of integrating and extending Next.js' functionality this system will enable.

## Proposal

To provide a seamless plugin experience, Next.js will add built-in support for auto-detecting packages with the naming convention `@next/plugin-`, `next-plugin-`, or `@<scope>/next-plugin-`. We will also investigate allowing local plugins in a `plugins` or `next-plugins` directory that don't require being installed.

Each of these plugins will have a meta field named `nextjs` in their `package.json` that gives us some extra information about the plugin like the required `env` values it relies on.

```json
// @next/plugin-google-analytics#package.json
{
  "name": "@next/plugin-google-analytics",
  "version": "9.1.2-canary.1",
  "nextjs": {
    "name": "Google Analytics",
    "required-env": ["GA_TRACKING_ID"]
  },
  "peerDependencies": {
    "next": "*"
  }
}
```

The only other required step for a plugin is to have a `src` directory with the lifecycle methods it wants to hook into.

To start, we explored what it would take to create a Google Analytics plugin and a Sentry plugin. Creating these example plugins gave us insight into the lifecycle methods that were needed.

## Currently exposed lifecycle methods

The naming scheme for lifecycle methods will be based on:

```
name + '-' + context
```

`name` will be based on the lifecycle, potentially prefixed by `on` if it's an event.

The `context` will be `client`, `server` or `build`. It's appended at the end of the file name so that if a lifecycle is available on both client and server it will be sorted together in editors.

Some of the initial methods that come to mind:

```js
// on-init-client.js
export default async function onInitClient({ router }): void {
  // called on client init, before hydration
}
```

```js
// on-error-client.js
export default async function onErrorClient({ err }): void {
  // called when an error occurs during render on the client
}
```

```js
// on-error-server.js
export default async function onErrorServer({ err }): void {
  // called when an error occurs during render on the server
}
```

```js
// document-head-tags-server.js
export default function headTags(DocumentCtx): Promise<ReactElement[]> {
  // called when rendering the initial document, allows adding tags to the head
  return (
    <>
      <script src="googleanalytics.com/code.js" async />
    </>
  );
}
```

```js
// document-body-tags-server.js
export default async function bodyTags(DocumentCtx): Promise<ReactElement[]> {
  // called when rendering the initial document, allows adding tags to the body
  return (
    <>
      <script src="example.com/body.js" async />
    </>
  );
}
```

```js
// babel-preset-build.js
module.exports = function babelPreset({ plugins, presets }, config): {
  // Babel config that will be merged inside next/babel
}
```

An example folder structure for one of these plugins would look like this

```
package.json
src/
  on-error-client.js
  on-error-server.js
  on-init-client.js
  on-init-server.js
```

Following this structure allows Next.js to provide a zero-config experience for plugins. Not only that, but it also allows us to code split these methods efficiently, only loading the parts of a plugin that are necessary at a specific point. We can also detect which parts of a plugin are being less efficient since they are all their own file and warn users of this.

## Providing information to plugins

Although we aim to make adding plugins to a project as simple as possible, some plugins still require some information like a `token` or an `id`.

To handle this, we allow you to provide these values using the current `env` config in `next.config.js` and the values will be injected into the plugin where needed during build time.

```js
// next.config.js
module.exports = {
  env: {
    GA_TRACKING_ID: "myTrackingId"
  }
};
```

If a value listed in a plugin's `required-env` is missing in your `next.config.js`'s `env` we will error and require you provide these values so that the plugin is able to function correctly. We may need to investigate name spacing these values to prevent collision between plugins.

**Comment by Janpot** [comment]
Interesting concept. In case you haven't seen it yet, [here's](https://marijnhaverbeke.nl/blog/extensibility.html) an interesting read about some considerations around plugin systems that I think apply here as well.

P.S. looks like there's a typo in the return type of
```js
export default function babelPreset(): Promise<ReactElement[]> {
```

**Comment by flybayer** [comment]
This'll be awesome!

We also need ability to **modify the webpack config**, for example to make `next-plugin-react-native-web`. 

Perhaps plugins can use the exact same `next.config.js` format for adding any modifications needed there (including things like `assetPrefix`).

**Comment by franklinjavier** [comment]
Love it 👊

**Comment by flybayer** [comment]
### What I need for `next-plugin-webextenson`

*Context: I am using Next.js to power a Chrome/Firefox extension*

I'm currently using a custom server during development to:
1. Write each page's html to disk on (a) server start and (b) page change
2. Copy all files in `public` to the same directory as the above pages

**Here's what I need:**

1. A hook into `app.renderToHTML()` that provides me with `req`, `res`, and output `html`
   - I'll use this to write that html to disk
2. A hook that provides me all the `pages` along with `app`
   - I'll use this on startup to call `app.renderToHTML()` for each page
   - This is essential because I can't wait for a client to request a page.
3. A hook on startup so I can copy all files in `public` to the same place as the page html files

**Comment by Jarred-Sumner** [comment]
This looks really cool!

> If a value listed in a plugin's required-env is missing in your next.config.js's env we will error and require you provide these values so that the plugin is able to function correctly.

I could see erroring when missing env vars being a source of frustration — sometimes, when working quickly, you install some packages but don’t end up using them. The next time the developer restarts their server in development, it would error. 

Perhaps instead, it could warn in development when the server first loads, skip loading the plugin, and then error during the production build? With a message that says to either provide the missing env vars or uninstall the plugin (with the package name, and ideally, a command to copy paste to uninstall)

Ideally, plugins would be able to optionally provide a label saying how to find the values for the env var, so that fixing the error is more clear.

**Comment by MANTENN** [comment]
I think it would be cleaner with hooks, and follow the hooks style. But again that's a matter of preference, and up to debate. Two options would be great.

**Comment by kachkaev** [comment]
> To provide a seamless plugin experience, Next.js will add built-in support for auto-detecting packages with the naming convention `@next/plugin-` and `next-plugin-`.

I'd also add autodetection for `@<scope>/next-plugin-` [like in Prettier](https://prettier.io/docs/en/plugins.html#using-plugins). This will be useful for company-scoped internal plugins.

With autodetection, I'm curious how to handle cases when order of plugin execution matters. E.g. one might want to make Sentry's `onErrorClient` / `onErrorServer` to be outermost even if the same methods are defined by other plugins 🤔 We might need some kind of '`z-index`' here, but as with CSS `z-index` [things can go a bit out of control one day](https://psuter.net/2019/07/07/z-index#fig-1) 😅

**Comment by thedankelly** [comment]
Can't wait.  👍🏻

**Comment by leerob** [comment]
I am _so_ excited about this 🎉

Let me know about the Sentry plugin - willing to help out!

**Comment by sanishkr** [comment]
Great idea. I will appreciate if a link full fledged sample next-plugin is provided. Also, any timelines on this?

**Comment by RichiCoder1** [comment]
I share the same concern about loading order, as well as for

> The only other required step for a plugin is to have a `src` directory with the lifecycle methods it wants to hook into.

It'd be nice if this could be configured for libraries that use transpilation so they can output to a separate folder.

Also it might be nice if a plugin could be organized as:
```
src/
  document/body-tags/server.js
  on-init/client.js
  on-init/server.js
```
though that may be overkill/perf problematic and a code smell for an overly large plugin.

> We will also investigate allowing local plugins in a `plugins` or `next-plugins` directory that don't require being installed.

I think this would be super helpful for developing and testing plugins without mucking with linking/`node_modules`

**Comment by ematipico** [comment]
It would be great if during the build stage, Next.js would be able to pass the *routes* of the application:
- posts/:postId
- contact-us 

Thanks!

**Comment by timneutkens** [comment]
> We also need ability to modify the webpack config, for example to make next-plugin-react-native-web.

We won't be exposing the webpack config to plugins as it's by far the easiest way to cause deoptimization. Instead we'll provide methods for aliasing modules for react-native-web for example.

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-543967469

Initially we won't be supporting these as they also sound like pitfalls for deoptimization.

For example:

> I'll use this to write that html to disk

Furthermore

> A hook that provides me all the pages along with app
I'll use this on startup to call app.renderToHTML() for each page
This is essential because I can't wait for a client to request a page.

This is not possible as in development there is [on-demand-entries](https://nextjs.org/blog/next-8#improved-on-demand-entries).

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-543968171

It's sort of dangerous to do so though, as it would be easy for you to forget providing environment variables in a certain environment and that would cause the plugin to either be excluded or error out based on the description.

Potentially we could do that behavior in development and fail if the env var is not set in production builds though.

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-543972592

> I think it would be cleaner with hooks, and follow the hooks style. But again that's a matter of preference, and up to debate. Two options would be great.

This comment is super confusing. Hooks has no relation to internal lifecycles (outside of React) which is what this RFC is about.

Could you elaborate what would be "cleaner". I'd argue it'd be **much** more confusing for users as these new methods have nothing to do with React directly.

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544021857

We already created a Sentry plugin to figure out which lifecycles are needed 👍

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544085262

As mentioned in other issues I generally don't share timelines. In general we release features incrementally starting with a minimal implementation and then iterate on that.

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544103648

> I share the same concern about loading order, as well as for

Loading order shouldn't matter in the majority of cases. Hence why we didn't include it in the initial proposal. One thing I've been discussing internally is having a bailout option in next.config.js called `plugins` where you'd be able to define order if you really have to.

> It'd be nice if this could be configured for libraries that use transpilation so they can output to a separate folder.

Plugins will go through the same compilation as other files in the project (pages etc). So they'll have the exact same configuration. One thing that we might want to keep in mind is TypeScript support, which should be easy to support.

Overall I'd strongly argue for not having plugin authors compile their code, it's very easy to deoptimize the code that way. Eg a wrong babel config would include  unnecessary polyfills and a wrong TypeScript config would inline helpers.

Instead Next.js will ensure the code is optimized and compiled correctly. This is especially important as module/nomodule support is coming soon.

> Also it might be nice if a plugin could be organized as:

That would make the structure much more complex for no real gain in my opinion.

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544104095

> It would be great if during the build stage, Next.js would be able to pass the routes of the application:

What would the use-case be.

> Edit: also, bear in mind that people usually put their final code inside a dist folder as they have plugins made using a different language (TypeScript). Does that mean that these plugins would need to compile their code into the same folder?

See my comment above 👍 https://github.com/zeit/next.js/issues/9133#issuecomment-544121540

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-543973456

> I'd also add autodetection for @<scope>/next-plugin- like in Prettier. This will be useful for company-scoped internal plugins.

Far point, we forgot to write that down in the proposal. I'll update it 💯

**Comment by epeli** [comment]
Hi!

This is very exciting! We're actually building a framework thing on top of Nextjs static export feature and currently we are completely wrapping Nextjs to be able run our things before and after Nextjs does its things. For now we just import the export function from `"next/dist/export"`.

So on top of my head here's few hooks we would need to create it as a plugin:

- `onNextDevStart`
  - We have a watcher that automatically compiles TypeScript types from *.graphql files as user edits them.
- `onExportStart`
  - Sending notification of export start. Would display it on a CMS admin panel
- `onPageExported`
  - We'd like to send exporting progress to an external service (eg. a CMS) so we could display exporting progress indicator there
 - `onExportDone`
    - For  generating sitemap.xml and robots.txt files to the exported output
    - Also for end notification like from `onExportStart`


Also could the plugins add #7208 like middleware?

**Comment by ematipico** [comment]
> What would the use-case be.

Serverless and lambdas.

I maintain a library that is like a bridge between terraform and next.js (and we have the same logic where I work). The routes are needed to configure the API Gateway. At the moment the routes are passed separately or calculated using the new next.js file structure. ([pageId].is).

So if next.js could provide the routes in a fashion manner, the integration with serverless services will be easier!

**Comment by Janpot** [comment]
> What would the use-case be.

I'd like to optimize my i18n. Instead of it being a React context with a store of messages (like `react-i18n` or so), I want to bake in messages at build time. For this to work I'd like to alias the `@locale` module (or another identifier) to a folder based on the path like `./locales/en` or `./locales/es`. That way I could generate localized components inside of them, and import like:

```js
import SomeMessage from '@locale/some-message';
```

In fact I'm only interested in 2 custom things in the build stage based on the path
- webpack alias (`resolve` option)
- setting `env`, to add `CURRENT_LANG` and `LANGUAGES` variables

**Comment by flybayer** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544118834

> Initially we won't be supporting these as they also sound like pitfalls for deoptimization.

I need the `app.renderToHTML()` hook in **development only**, so how would that be a pitfall for deoptimization? 

I updated my original comment to include the following:

> The absolute minimum I need is:
> 
> 1. A development-only hook into `app.renderToHTML()` that provides me with `req`, `res`, and output `html`
>    - I'll use this to write that html to disk
> 2. A development-only hook after server boot that provides `app`
>    - I'll use this to do everything else I need like copy certain files, etc.

**Comment by rauchg** [comment]
Not sure if this has been contemplated, but I think it's really important to show what plugins are enabled at bootup time:

```
✓ Loaded plugin next-plugin-styled-components
✓ Loaded plugin next-plugin-google-analytics
```

**Comment by eweilow** [comment]
Would it be possible to have access to the request (and if it makes sense, the response as well) for server-side errors? 

```typescript
// on-error-server.js
export default async function onErrorServer({ err, req }): void {
  // can now log what URL throws specific errors
}
```

Including other information, such as the current page and the query, could be useful to get better context for error logging too.

**Comment by franky47** [comment]
> Not sure if this has been contemplated, but I think it's really important to show what plugins are enabled at bootup time:

That would be very useful indeed. What about also adding their respective version, which at some point might become as important as the name of the plugin itself (for debugging and security) ?
```
✓ Loaded plugin next-plugin-styled-components@1.2.3
✓ Loaded plugin next-plugin-google-analytics@4.5.6
```

The `@` separator (used by NPM and Yarn in their output) makes it harder to distinguish the plugin name from the version (especially across multiple lines where the separator is not aligned vertically).

A tabular output could solve this issue and could add inclusion error details as per https://github.com/zeit/next.js/issues/9133#issuecomment-544118998, but it would be harder to parse by external tools.

**Comment by borekb** [comment]
Awesome to see this RFC! First impressions:

I was surprised that plugins are "activated" just by being installed. One consequence is the necessity to juggle with environment variables, be careful to avoid name clashes, etc., instead of doing something like this:

```js
// next.config.js
module.exports = {
  plugins: [
    {
      plugin: 'nextjs-google-analytics',
      options: {
        trackingId: '...'
      }
    }    
  ]
};
```

BTW, for in-development plugins, I'd probably like to use something like this:

```js
plugin: '../my-plugins/nextjs-google-analytics',
```

I.e., standard Node module resolution.

I didn't quite understand the automatic resolution from `src`. In TypeScript projects, it's uncommon to transpile from `src` to `src` – much more common is `dist` or `lib`, so I'm not sure why to resolve from `src`, specifically.

Maybe it's related to the file-based lifecycle convention? That was another thing that caught my attention – it's uncommon to require many files, each with a single `default` export, vs. just a normal structure with specifically named functions. It makes code splitting easier but is that really so important for plugins? I feel some developer ergonomics have been sacrificed here.

That's probably all at this point, I'll let it sink in a bit more 😄.

**Comment by franky47** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544121540
> Overall I'd strongly argue for not having plugin authors compile their code, it's very easy to deoptimize the code that way. Eg a wrong babel config would include unnecessary polyfills and a wrong TypeScript config would inline helpers.

What would that imply for plugin authoring in TypeScript, especially with regards to importing Next.js types and exporting plugin types to applications using it ? Would the TS configuration be the one from Next.js ?

I've been in a situation recently where I had to have two separate `tsconfig.json` files, one for Next.js and one for the server code (also written in TypeScript and ran with `ts-node`). I'd love to see zero-config TypeScript support for all parts of a Next.js application, including plugins.

**Comment by jacobrask** [comment]
The way we have started delivering Next plugins internally is together with the associated npm package, for example @org/polyfill/next-plugin.js and we could do the same with our UI library

Just something to consider as a workflow that others might also be using but is not visible in open source.

**Comment by timneutkens** [comment]
> https://github.com/zeit/next.js/issues/9133#issuecomment-544202885
> https://github.com/zeit/next.js/issues/9133#issuecomment-544203720

@borekb @franky47 it seems like you're both misunderstanding, you **shouldn't** compile plugins yourself. As said here:

https://github.com/zeit/next.js/issues/9133#issuecomment-544121540

> Plugins will go through the same compilation as other files in the project (pages etc). So they'll have the exact same configuration. One thing that we might want to keep in mind is TypeScript support, which should be easy to support.

> Overall I'd strongly argue for not having plugin authors compile their code, it's very easy to deoptimize the code that way. Eg a wrong babel config would include unnecessary polyfills and a wrong TypeScript config would inline helpers.


With regards to:

> especially with regards to importing Next.js types and exporting plugin types to applications using it ?

As of the current proposal user code is not going to access plugin code directly. So there are no "plugin types" to be exported.

**Comment by timneutkens** [comment]
> I was surprised that plugins are "activated" just by being installed. One consequence is the necessity to juggle with environment variables, be careful to avoid name clashes, etc., instead of doing something like this:

One thing to keep in mind is that many of the lifecycles get bundled into browser code. Meaning that they should be *as small as possible*. So the environment approach works well since they variables are inlined at build time. Meaning you get the smallest possible access to variables. Furthermore `env` is restricted in what can be passed through so people won't try to serialize massive objects accidentally.

**Comment by borekb** [comment]
@timneutkens thanks for the reply. I'm no expert on code splitting and feel like several design points are guided by it, could you help me understand more?

1. How important is code splitting between different contexts (client vs. server) vs. code splitting within a client-side / server-side contexts? For example, I can imagine that it's _very_ important to send only client-side code to the browser but is it as important to code-split between client-side functions, for example?
2. Are environment variables the only way to inline values? I understand that it's convenient since [build-time env vars are already inlined](https://github.com/zeit/next.js/tree/d94d2700f945b68166f2defdd4e35a8733dc1b12#exposing-configuration-to-the-server--client-side) but if you had, say, more strongly-typed `options` API, would it be possible to inline it via webpack as well?
3. If, hypothetically, Next.js plugins were normal Node modules – folder with `package.json` that contains the `main` field which points to a `.js` file that exports a couple of functions – would you be also able to code-split it if dynamic imports were used? I have never delved too deep into them so sorry for my ignorance.

What I'm trying to understand is whether code-splitting concerns (which are generally very important!) necessarily lead to the proposed structure, or if many of the same benefits could be achieve with "normal JS/TS code" that has a `.d.ts` shipped for it, I get full VSCode support, etc.

**Comment by timneutkens** [comment]
@borekb 

1. One example is when a module is not included in the app so the lifecycle isn't needed at all in the particular app. We don't want to bundle code for it that isn't needed.
2. It's more so that we don't want people to introduce giant configuration objects that are sent to the client-side as this would cause bundle sizes to grow considerably.
3. If plugins were a single file maybe, but they're not. Also keep in mind that as a user of a plugin you're not going to `import something from 'next-plugin-example`. The lifecycles are going to be hooking into the system automatically. I feel like this is the main point where you're confused. It basically isn't needed to add a `.d.ts` for the plugin itself as there is nothing to be imported from user code.

**Comment by timneutkens** [comment]
@jacobrask 
> One way we have experimented with delivering Next plugins internally is together with the associated npm package, for example @org/polyfill/next-plugin.js and we could do the same with our UI library

This sounds interesting, however it's also tricky in so far that it could be that the design system uses styled-component (just an example) and another plugin is added for styled-components too, causing multiple issues with rendering / styles hydration.

It did make me think that eg `material-ui` would be able to automatically configure Next.js without you having to add `next-plugin-material-ui` or similar. However that'd suffer from the same issue of material-ui using Emotion and if your app is also using emotion separately it'd cause issues too.

**Comment by borekb** [comment]
@timneutkens 

1. Got it, thanks 👍.
2. Looking at Gatsby plugins, for example, I don't think people are creating huge configs. But OK, it is _some_ concern, I agree.
3. I think I'm not confused – I was referring to a use case when I'm writing the plugin, not consuming it. Your API is file-based instead of function-based which is "strange" and I'm looking for reasons why it's that way. Again, if plugins were single-file + dynamic imports, what would be the problem?

**Comment by timneutkens** [comment]
2. We exactly don't want that approach of having a single file.
3. It's extremely easy to deoptimize if you don't. For example if it's a single file defining a variable in the module scope would already cause tree shaking for a module used to be opted out of. Instead with the separate file approach it's increasingly harder to deoptimize it.

**Comment by rlindskog** [comment]
@epeli 
> BTW, for in-development plugins, I'd probably like to use something like this:
> 
> ```js
> plugin: '../my-plugins/nextjs-google-analytics',
> ```
> 
> I.e., standard Node module resolution.

It might be cool to automatically resolve "in-development" plugins - maybe by detecting a plugins folder?

Ex:
```
package.json
pages
  index.js
plugins
  styled-components
    on-error-client.js
    on-error-server.js
    on-init-client.js
    on-init-server.js 
```

**Comment by karlhorky** [comment]
> @borekb: I was surprised that plugins are "activated" just by being installed.

Yeah, for me this is both cool (eg. I can understand the benefit) and also surprising (eg. maybe it could cause weird problems or not allow enough control). 

For an example of "not enough control", I think about this comment from @kachkaev about situations where execution order matters (and I think this will not be uncommon to come up):

> With autodetection, I'm curious how to handle cases when order of plugin execution matters

Another point that I just thought of was about transitive dependencies. For instance, consider:

1. You install an unrelated dependency X that has a dependency Y on something that has been compromised similar to [the `event-stream` compromise](https://snyk.io/blog/malicious-code-found-in-npm-package-event-stream/)
2. The compromise in dependency Y is that it installs a Next.js plugin that does something malicious to the site
3. You do not notice that a Next.js plugin has been installed and you deploy your site to production
4. Something bad happens 💥

But maybe transitive dependencies won't be auto-loaded?

---

Either way, it feels like having an **ordered** `plugins` array in the Next.js config would be a less magic, more careful and controlled way of adding plugins.

Like Gatsby does it, for example: https://www.gatsbyjs.org/docs/gatsby-config/#plugins

**Comment by Janpot** [comment]
> Either way, it feels like having an **ordered** `plugins` array in the Next.js config would be a less magic, more careful and controlled way of adding plugins.

It could also serve as a place to add configuration, instead of loose variables in the `env` key.

**Comment by timneutkens** [comment]
@karlhorky The argument doesn't hold true because the loading is backed on `package.json` in your project, sub-deps wouldnt' be loaded.

**Comment by karlhorky** [comment]
> The security argument doesn't hold true because the loading is backed by package.json in your project, sub-deps wouldnt' be loaded.

Yep, expected as much. 👍

So it would be relevant in fewer cases then, such as when users install something accidentally (for example, a package with a long install string, where the user doesn't check which packages are in it), but these may prove to be acceptable levels of risk.

> I'm not opposed to have a bailout method through a `plugins` key if you really want to opt-out of automatic loading though.

This sounds like a reasonable compromise: zero-config by default, but if you want control, you can opt in.

**Comment by timneutkens** [comment]
> So it would be relevant in fewer cases then, such as when users install something accidentally (for example, a package with a long install string, where the user doesn't check which packages are in it), but these may prove to be acceptable levels of risk.

Sounds strange to me. Especially this part:

> such as when users install something accidentally (for example, a package with a long install string, where the user doesn't check which packages are in it)

This sounds very unlikely and I'd argue that at that point it doesn't really matter if you're using Next.js or not, your laptop would already be compromised before of pre-install / post-install scripts with any dependency.

**Comment by karlhorky** [comment]
Yep, I agree that this is strange behavior. But I've seen it a lot from developers in the wild. Shouldn't pass code review though, so hopefully not as present as it could be.

The difference between pre-/post-install risk and auto-loading packages is that auto-loading loads code many more times across more environments, whereas pre-/post-install scripts load code once at install time.

The contract of **having to import packages** to trigger running the code contained in them **every time your application starts** can be seen as a security measure, one that would no longer exist with auto-loading.

But again, like I said, maybe this is an acceptable level of risk.

**Comment by evantahler** [comment]
Just chiming in to say that I'm I'm very excited about this and happy to help! 
We've been monkey patching a way to let plugins add their own pages for a while now, as described here https://spectrum.chat/next-js/general/creating-an-application-with-front-end-plugins~ed8926da-62bf-438d-8916-172900b3e9bb ... and it's less than optimal :D

**Comment by jth0024** [comment]
I just want to chime in that ordering of plugins will be very important. I actually built a lightweight plugin system on top of NextJS for my company. We initially wanted to auto load plugins from a directory to match NextJS’s convention over configuration approach, but we quickly ran into issues. 1) plugins will need data passed to them, and 2) order matters. It’s very common that other plugins might depend on one loaded before them. Likewise, it’s common that you might depend on one plugin for the initialization of another. For example, we had a config provider plugin that was needed to read config values that were passed into a redux plugin. Just some food for thought! Don’t hesitate to reach out if you want some more info about the solution we implemented. Very excited about this!

**Comment by HaNdTriX** [comment]
## Plugin Scope

Next.js transitioned in the last version to a more page based approach.
Instead of installing things on global (`_app.js`) scope we prefer now a local (page centric) scope.

Examples are the new [with-apollo](https://github.com/zeit/next.js/tree/canary/examples/with-apollo) or [with-redux](https://github.com/zeit/next.js/tree/canary/examples/with-redux) examples.

How do plugins fit into this context. Are plugins always global (enabled project wide) or will it be possible do enable/disable plugins on page level?

**Comment by satazor** [comment]
> We won't be exposing the webpack config to plugins as it's by far the easiest way to cause deoptimization. Instead we'll provide methods for aliasing modules for react-native-web for example.

@timneutkens This will limit the plugins usefulness. Currently, Next.js already provides a way to change the webpack config via `webpack()` in next.config.js. In fact, the "unofficial" way plugins are set right now is to compose various packages in your `next.config.js`, where all these packages can already modify the webpack config.

I understand you are trying to be protective by limiting people from doing bad things, but you are also limiting experienced people from doing great things.

**Comment by satazor** [comment]
I would also would like to say that this is great, and will make the ecosystem healthier. From what I could evaluate, Gatsby plugin system is great and we could take some inspiration on it.

**Comment by jkrems** [comment]
One use case I haven't heard mention yet: (API) routes provided by a plugin. Examples could be a GraphQL endpoint or (in my case) dynamically generated scripts. Right now that requires creating a file in `pages/` even if that file is just a one-liner re-exporting the plugin implementation of a request handler.

**Comment by brad-decker** [comment]
@timneutkens this proposal excites me, would allow us to start tearing down our custom _document.js file which has grown into a monster. I see that you're already starting to test this with the material-ui / google analytics plugin. Looks like exciting stuff. Right now we're doing styled-components + material-ui in _document and it looks pretty horrific. With a plugin system we'd be able to separate those into two different local plugins that'd be pretty great. 

:+1:

**Comment by timneutkens** [comment]
@brad-decker tbh I don't feel like that would be super great, it means you're already using 2 css-in-js solutions in your app.

**Comment by brad-decker** [comment]
@timneutkens we are using the material ui core components but are using styled-components as our preferred method of extending those styles. Material-ui will be moving its base engine to styled-components "soon" https://material-ui.com/guides/interoperability/ but i do agree that it's worrisome. Are there any gotchas i should be worried about with this setup from the next perspective?

**Comment by jacobrask** [comment]
In production apps you are likely to at least at some point (during a migration) use more than one css-in-js solution, unless everyone selects the right library the first time and never switch.

**Comment by brad-decker** [comment]
@jacobrask -- true, but this feels a little ... bad.

```
 ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: App => props =>
            sheet.collectStyles(materialSheets.collect(<App {...props} />)),
        });

```

**Comment by samjbmason** [comment]
I can see it was mentioned in the initial outline but will there be a way to develop plugins locally without the need to creating a npm package and linking it?

**Comment by Ianfeather** [comment]
👋 Hi there, is there a timeline for this landing?

**Comment by hugo** [comment]
> I can see it was mentioned in the initial outline but will there be a way to develop plugins locally without the need to creating a npm package and linking it?

You can do this by using a [local path](https://docs.npmjs.com/files/package.json#local-paths) for the package. For example I have made a plugin for adding a `polyfill.io` `<script>` tag to every page. The package is in `packages/next-plugin-polyfill` of my project directory. In my `package.json` I have (other details elided):
```json
{
  "dependencies": {
    "next-plugin-polyfill": "file:plugins/next-plugin-polyfill"
  }
}
```

When you run `npm install` (or `npm ci`) `npm` will symlink your local directory into `node_modules`.

**Comment by bleonard** [comment]
A thing we are struggling with is how plugins could provide components. For example, let's say we made a page that was able to load content from other places. There would be a certain interface you'd register for.: "I know how to import stuff from Slack" and "I know how to import stuff from s3" and so on. The main next page doesn't `import` or `require` these ahead of time, but asks "who are all my importers" from a central registry. Then gives the user a choice. Maybe it shows an icon that each plugin specifies. Then when the icon is clicked, it shows a component contributed by the plugin in a modal, passing it a callback. Maybe the slack one knows how to browse rooms to find a message, for example. Upon "success" as defined by the UI component from the plugin, it calls the callback to load the content into the main thing. Ideally, all of this happens by including N plugins in my `package.json` file, but it's unclear to me how to do that without the explicit `import` statements.

**Comment by adamsoffer** [comment]
Really excited about this. Would a nextjs plug-in system unlock data sourcing ala gatsby? I’d LOVE to be able to install a plugin and be able to query graphql data out of the box from all the awesome headless cms tools out there. https://www.gatsbyjs.org/plugins/?=gatsby-source

**Comment by martpie** [comment]
> To handle this, we allow you to provide these values using the current env config in next.config.js and the values will be injected into the plugin where needed during build time.

I would recommend adding a way to do that at runtime rather than build time. As soon as you use a solution like Docker to package and distribute your app, you want to change the behavior of your app with runtime environment variables, rather than build a different Docker image for each option you change.

**Comment by clRcoDE** [comment]
cant wait for it 
this will makes developers life a lot easier with @official-plugins , automated webpack configs and (_document , _app ) changes 
nextjs & zeit are going to be my primary tools

**Comment by eric-burel** [comment]
Hi, this is probably more a "next step" but something you'd want to consider for plugins is UI. That probably goes beyond the idea of "plugins" to form fully independent "packages".

An example: say I want to implement a debug package, with an admin interface. You'd like it to add pages available for admin users.

That could mean having a "prebuild" or "install" step for plugins, that would allow for more complex file manipulation like adding pages into the users project.

**Comment by eric-burel** [comment]
Here is a link to a more detailed reference to what we want to implement in Vulcan with Next: https://github.com/VulcanJS/vulcan-next-starter/issues/3

The issue might not be very clear to non-Meteor/non-Vulcan devs but basically what we try to do is to bring the power of Meteor isomorphic packages and Vulcan dynamic generation (Mongo collection, graphql API, routes and overrideable React components) to what we could call the "static" ecosystem, through Next. 

A powerful plugin system in Next means an easier transition for us.

To sum it up, our plugins plugin should be able:
- to create new routes
- to define code that should run at server startup (for example to define Mongoose schema)
- to generate static files in the build folder
- to extend whatever can be extended in Next: _app, _document, pages, custom server and so on.

**Comment by fillipvt** [comment]
I'm wondering the same as @HaNdTriX and would like to understand how this would fit within a page-centered approach.

Would there ever be a need for a plugin to run or be used on certain pages and not others? And if yes, how would this be supported within the context of this RFC?

**Comment by chrisrichard** [comment]
> This is very exciting! We're actually building a framework thing on top of Nextjs static export feature and currently we are completely wrapping Nextjs to be able run our things before and after Nextjs does its things. For now we just import the export function from `"next/dist/export"`.

:+1:

I've hacked in something in a fork to enable NewRelic monitoring of the export process (https://github.com/ratehub/next.js/commit/9a46ba7694bd2212b5f1d127a5424b60f7797e56) and would be great to be able to do this with hooks.

**Comment by smakosh** [comment]
Does this mean that all the next js templates can be turned into plugins we install and when there is a new release, all we have to do is upgrade that npm package?
